### PR TITLE
check  master server for mooncake store

### PR DESCRIPTION
**Pull request by huangtingwei9988**
## Motivation
When deploying Mooncake Store online, the startup order of Mooncake Store master and sglang may not be guaranteed. Therefore, a check server is added. When Mooncake Store Master has not yet been fully started successfully, sglang waits for the master for 10 minutes to prevent rapid failure when executing `store.setup`

cc @ykwd @stmatengss 

co-author @zhaoyongke
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py [hunk 1]
```diff
@@ -171,6 +177,52 @@ def __init__(self, storage_config: HiCacheStorageConfig = None):
...
+
+            if check_health_resp.text != "OK":
+                logger.info("waiting Mooncake store master started, cost_time: %s", time.time() - start_time)
+                time.sleep(3)
+                continue
+
+            health_result = True
+
+            try:
+                check_segments_resp = requests.get(segments_url)
```
Comment (stmatengss) [inline python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py]: Why need to check all segments?

File: python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py [hunk 2]
```diff
@@ -1,16 +1,19 @@
...
 from typing import Any, List, Optional
+import requests
 
 import torch
 
 from sglang.srt.mem_cache.hicache_storage import HiCacheStorage, HiCacheStorageConfig
 
 DEFAULT_GLOBAL_SEGMENT_SIZE = 4 * 1024 * 1024 * 1024  # 4 GiB
 DEFAULT_LOCAL_BUFFER_SIZE = 16 * 1024 * 1024  # 16 MB
+SETUP_TIMEOUT = 600 # 10min
```
Comment (stmatengss) [inline python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py]: Ten minutes is too long to start Mooncake Master. I guess 60 seconds might be preferable.

File: python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py (hunk 1 above)
Comment (stmatengss) [inline python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py]: The logic should be like this:
```
While the interval in setup_timeout
    if check(health):
        if check(check_segments):
             is_success = true
             break
```

**Comment by ykwd** [comment]
It’s a good idea to prevent startup failures caused by the master server starting late. However, I have some concerns about the current solution:

* The HTTP server port used to query master information is hard-coded as `9003`. This is problematic because the port number is configurable when starting the HTTP server.
* Potential deadlocks may occur. The SGLang server waits for the master to become active and for at least one global segment to be registered before it creates the store client and registers its own segment. But if users choose not to run a separate store service—instead relying entirely on segments provided by the SGLang server—this could lead to deadlocks.
* The `check_server` function sends HTTP requests directly to `self.config.master_server_address`. However, in HA mode, `master_server_address` actually points to the etcd endpoints. In this case, the Mooncake client first queries etcd through this address to obtain the actual master server address.

For these reasons, I think it would be better to implement the waiting logic inside the Mooncake store client. Whether or not to wait for the master to start could then be exposed as an optional client parameter.

**Comment by stmatengss** [comment]
Checking segments is enough to ensure a healthy status. It can simplify the implementations.

File: python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py (hunk 2 above)
Comment (zhaoyongke) [reply python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py]: In our case, init a huge segment(TB level) will take minutes, 60s will always timeout

File: python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py (hunk 1 above)
Comment (zhaoyongke) [reply python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py]: > Why need to check all segments?

Warmup will put a key-value into store, so at least a segment should be mounted before warmup.

**Comment by zhaoyongke** [comment]
> * The HTTP server port used to query master information is hard-coded as `9003`. This is problematic because the port number is configurable when starting the HTTP server.

Set a unified env variable for both sglang and mooncake config?

> * Potential deadlocks may occur. The SGLang server waits for the master to become active and for at least one global segment to be registered before it creates the store client and registers its own segment. But if users choose not to run a separate store service—instead relying entirely on segments provided by the SGLang server—this could lead to deadlocks.

Exactly. We should put self.check_server() after ret_code = self.store.setup(...) in this case

> * The `check_server` function sends HTTP requests directly to `self.config.master_server_address`. However, in HA mode, `master_server_address` actually points to the etcd endpoints. In this case, the Mooncake client first queries etcd through this address to obtain the actual master server address.

Known issue. Now only support non-HA mode.

**Comment by ykwd** [comment]
> > * The HTTP server port used to query master information is hard-coded as `9003`. This is problematic because the port number is configurable when starting the HTTP server.
> 
> Set a unified env variable for both sglang and mooncake config?

Mooncake master already supports configuring this parameter. Any way to allow configuring this in sglang side should work.
> 
> > * Potential deadlocks may occur. The SGLang server waits for the master to become active and for at least one global segment to be registered before it creates the store client and registers its own segment. But if users choose not to run a separate store service—instead relying entirely on segments provided by the SGLang server—this could lead to deadlocks.
> 
> Exactly. We should put self.check_server() after ret_code = self.store.setup(...) in this case

`check_server` is called before `store.setup` to ensure the success of the latter. Putting it after `store.setup` may not help with this purpose.

> 
> > * The `check_server` function sends HTTP requests directly to `self.config.master_server_address`. However, in HA mode, `master_server_address` actually points to the etcd endpoints. In this case, the Mooncake client first queries etcd through this address to obtain the actual master server address.
> 
> Known issue. Now only support non-HA mode.

Currently, both HA and Non-HA mode are available to users. If we only support non-HA mode, then HA mode users will no longer be able to start the mooncake backend, as `check_server` will not pass under non-HA mode.

To ensure compatibility with HA mode and to make the “wait-for-master” logic easier to maintain, I recommend implementing it on the Mooncake client side. If we really want to handle this on the SGLang side, then at the very least, it should be made an optional feature.

**Review by stmatengss — APPROVED** [review]
LGTM

**Review by stmatengss — CHANGES_REQUESTED** [review]
It seems we can't pass the test in local run.

**Comment by huangtingwei9988** [comment]
> It seems we can't pass the test in local run.

@stmatengss After discussing with @ykwd, I added an option to open it if necessary.

**Review by stmatengss — APPROVED** [review]

**Comment by stmatengss** [comment]
@ykwd Could you double check this changes？

**Review by hzh0425 — APPROVED** [review]

File: python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py [hunk 3]
```diff
@@ -63,10 +71,16 @@ def from_file() -> "MooncakeStoreConfig":
                 config.get("global_segment_size", DEFAULT_GLOBAL_SEGMENT_SIZE)
             ),
             # Zero copy interface does not need local buffer
-            local_buffer_size=DEFAULT_LOCAL_BUFFER_SIZE,
+            local_buffer_size=config.get(
+                "local_buffer_size", DEFAULT_LOCAL_BUFFER_SIZE
```
Comment (ykwd) [inline python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py]: Why do we need local_buffer_size parameter?

Comment (zhaoyongke) [reply python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py]: Keep this to be compatible with non-zerocopy calls, such as warmup().

Comment (ykwd) [reply python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py]: Currently, the default 16 MB setting should be sufficient for the warmup, and there are no other parts of the code that use the local buffer.
Perhaps we could consider hiding this parameter to avoid confusing users.

Comment (huangtingwei9988) [reply python/sglang/srt/mem_cache/storage/mooncake_store/mooncake_store.py]: okk @ykwd

**Review by xiezhq-hermann — APPROVED** [review]
